File: MergeSort/swift/src/binary_search.swift
/// Performs binary search to find the index of target in a sorted array.
///
/// - Parameters:
///   - arr: A sorted array of integers
///   - target: The value to search for
/// - Returns: The index of target if found, nil otherwise
///
/// - Complexity:
///   - Time: O(log n)
///   - Space: O(1)
func binarySearch<T: Comparable>(_ arr: [T], target: T) -> Int? {
    guard !arr.isEmpty else {
        return nil
    }

    var left = 0
    var right = arr.count - 1

    while left <= right {
        let mid = (left + right) / 2

        if arr[mid] == target {
            return mid
        } else if arr[mid] < target {
            left = mid + 1
        } else {
            if mid == 0 {
                break
            }
            right = mid - 1
        }
    }

    return nil
}

// Example usage
func main() {
    // Test cases
    let testCases: [(arr: [Int], target: Int)] = [
        ([1, 2, 3, 4, 5], 3),    // Target in middle
        ([1, 2, 3, 4, 5], 1),    // Target at start
        ([1, 2, 3, 4, 5], 5),    // Target at end
        ([1, 2, 3, 4, 5], 6),    // Target not present
        ([], 1),                  // Empty array
        ([1], 1)                  // Single element
    ]

    for (arr, target) in testCases {
        let result = binarySearch(arr, target: target)
        print("Array: \(arr), Target: \(target), Result: \(String(describing: result))")
    }

    // Test with strings
    let stringTestCases: [(arr: [String], target: String)] = [
        (["apple", "banana", "cherry", "date", "elderberry"], "cherry"),
        (["apple", "banana", "cherry", "date", "elderberry"], "apple"),
        (["apple", "banana", "cherry", "date", "elderberry"], "elderberry")
    ]

    print("\nTesting with strings:")
    for (arr, target) in stringTestCases {
        let result = binarySearch(arr, target: target)
        print("Array: \(arr), Target: \(target), Result: \(String(describing: result))")
    }
}

main()

File: MergeSort/rust/src/binary_search.rs
/// Performs binary search to find the index of target in a sorted array.
///
/// # Arguments
///
/// * `arr` - A sorted slice of integers
/// * `target` - The value to search for
///
/// # Returns
///
/// * `Option<usize>` - The index of target if found, None otherwise
///
/// # Examples
///
/// ```
/// let arr = vec![1, 2, 3, 4, 5];
/// assert_eq!(binary_search(&arr, 3), Some(2));
/// assert_eq!(binary_search(&arr, 6), None);
/// ```
///
/// # Time Complexity
/// O(log n)
///
/// # Space Complexity
/// O(1)
pub fn binary_search(arr: &[i32], target: i32) -> Option<usize> {
    if arr.is_empty() {
        return None;
    }

    let mut left = 0;
    let mut right = arr.len() - 1;

    while left <= right {
        let mid = (left + right) / 2;

        match arr[mid].cmp(&target) {
            std::cmp::Ordering::Equal => return Some(mid),
            std::cmp::Ordering::Less => left = mid + 1,
            std::cmp::Ordering::Greater => {
                if mid == 0 {
                    break;
                }
                right = mid - 1;
            }
        }
    }

    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_binary_search() {
        // Test with target in middle
        assert_eq!(binary_search(&[1, 2, 3, 4, 5], 3), Some(2));

        // Test with target at start
        assert_eq!(binary_search(&[1, 2, 3, 4, 5], 1), Some(0));

        // Test with target at end
        assert_eq!(binary_search(&[1, 2, 3, 4, 5], 5), Some(4));

        // Test with target not present
        assert_eq!(binary_search(&[1, 2, 3, 4, 5], 6), None);

        // Test with empty array
        assert_eq!(binary_search(&[], 1), None);

        // Test with single element
        assert_eq!(binary_search(&[1], 1), Some(0));

        // Test with negative numbers
        assert_eq!(binary_search(&[-5, -3, -1, 0, 2], -3), Some(1));

        // Test with duplicate elements (should return first occurrence)
        assert_eq!(binary_search(&[1, 2, 2, 2, 3], 2), Some(1));
    }
}

fn main() {
    // Example usage
    let test_cases = vec![
        (vec![1, 2, 3, 4, 5], 3),      // Target in middle
        (vec![1, 2, 3, 4, 5], 1),      // Target at start
        (vec![1, 2, 3, 4, 5], 5),      // Target at end
        (vec![1, 2, 3, 4, 5], 6),      // Target not present
        (vec![], 1),                    // Empty array
        (vec![1], 1),                   // Single element
    ];

    for (arr, target) in test_cases {
        let result = binary_search(&arr, target);
        println!("Array: {:?}, Target: {}, Result: {:?}", arr, target, result);
    }
}
